core-util/v2/impl/FP0.hpp

```diff
@@ -0,0 +1,238 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef __FUNCTIONPOINTER_IMPL_FP0_HPP__
+#define __FUNCTIONPOINTER_IMPL_FP0_HPP__
+#ifndef __FUNCTIONPOINTER_HPP__
+#error FP0.hpp is an internal file that should only be called from inside FunctionPointer.hpp
+#endif
+
+template<typename R, size_t I>
+class FunctionPointer<R(), I> {
+public:
+    typedef struct arg_struct{
+    } ArgStruct;
+    typedef R(*static_fp)(void);
+
+    FunctionPointer(R (*fp)()) {
+        new(_storage) impl::StaticPointer<R(), I>(fp);
+    }
+    template <typename C>
+    FunctionPointer(C *c, R (C::*member)()) {
+        new(_storage) impl::MethodPointer<R(), C, I>(c, member);
+    }
+    template <typename F>
+    FunctionPointer(const F & f) {
+        new(_storage) impl::FunctorPointer<R(), F, I>(f);
+    }
+    operator bool() const {
+        return *reinterpret_cast<impl::FunctionPointerInterface<R()> *>(&_storage);
+    }
+    bool operator ==(const FunctionPointer<R()> & rhs) const {
+        return *reinterpret_cast<impl::FunctionPointerInterface<R()> *>(&_storage) ==
```

Comment:
Using external vtable instead of inheritance, it is possible to make a valid operator== but the complexity will increase by an order of magnitude for peoples non initiated. 

(
Another advantage of external vtable is the reduction of template instantiation: 
Today, each function pointer instantiate the following functions:
  - virtual operator bool() const = 0;
  - virtual bool operator ==(const FunctionPointerInterface &) const = 0;
  - virtual R call(void) const = 0;
  - virtual ~FunctionPointerInterface(){};
  - virtual void copy_to(void *) const = 0;

With an external vtable, we can control more finely template instantiations and it became possible to instantiate:
  - `virtual operator bool() const = 0` only one instantiation for all POD
  - `virtual bool operator ==(const FunctionPointerInterface &) const` only one instance for all POD
  - `virtual ~FunctionPointerInterface()` only one instance for all POD
  - `virtual void copy_to(void *) const` only one instance for all POD
)